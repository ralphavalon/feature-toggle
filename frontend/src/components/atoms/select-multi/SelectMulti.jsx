import React from 'react';
import CreatableSelect from 'react-select/creatable';
import PropTypes from 'prop-types';

const getValueAs = (value, type) => {
  let response = [];

  if (value.length > 0) {
    for (let i in value) {
      if(type === 'object') {
        response.push({ value: value[i], label: value[i] });
      } else {
        response.push(value[i].value);
      }
    }
  }
  return response;
};

const showValue = value => {
  if (!value) return null;
  return value.length > 0 && typeof value[0] === 'string'
    ? getValueAs(value, 'object') : value;
};

const toPlanList = value => {
  if (!value) return null;
  return value.length > 0 && typeof value[0].value === 'string'
    ? getValueAs(value, 'string') : value;
};

const SelectMulti = ({
  name, value, options, onChange, placeholder, disabled
}) => {

  return (
    <CreatableSelect
      isMulti
      name={name}
      value={showValue(value)}
      onChange={(option) => {
        onChange(toPlanList(option));
      }}
      isDisabled={disabled}
      placeholder={placeholder}
      options={showValue(options)}
    />
  );
};

SelectMulti.propTypes = {
  name: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  options: PropTypes.array,
  placeholder: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
};

export default SelectMulti;