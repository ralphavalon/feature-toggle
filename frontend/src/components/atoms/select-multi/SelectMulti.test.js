import React from 'react';
import { shallow } from 'enzyme';

import SelectMulti from './SelectMulti';

afterEach(() => {
  jest.clearAllMocks();
});

const selectMultiRequiredProps = {
  name: 'name'
};

let selectMultiOptionalProps = {
  disabled: true,
  options: ['123'],
  placeholder: 'placeholder',
  value: ['123']
};

describe('SelectMulti', () => {
  it('renders correctly with required props', () => {
    const wrapper = shallow(<SelectMulti {...selectMultiRequiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });
  
  it('renders correctly with required props and optional props', () => {
    const wrapper = shallow(<SelectMulti {...{...selectMultiRequiredProps, ...selectMultiOptionalProps}} />);
    expect(wrapper).toMatchSnapshot();
  });
})

