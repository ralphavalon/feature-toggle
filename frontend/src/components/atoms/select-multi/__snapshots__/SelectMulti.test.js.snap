// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectMulti renders correctly with required props 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isMulti={true}
  name="name"
  onChange={[Function]}
  options={null}
  value={null}
/>
`;

exports[`SelectMulti renders correctly with required props and optional props 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={true}
  isMulti={true}
  name="name"
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "123",
        "value": "123",
      },
    ]
  }
  placeholder="placeholder"
  value={
    Array [
      Object {
        "label": "123",
        "value": "123",
      },
    ]
  }
/>
`;
